composition-js/src/__tests__/compose.custom.test.ts

```diff
@@ -0,0 +1,496 @@
+    }
+  });
+
+  it('type-system directive, core feature, not exposed', () => {
+    const subgraphA = {
+      name: 'subgraphA',
+      typeDefs: gql`
+      extend schema
+        @link(url: "https://specs.apollo.dev/link/v1.0")
+        @link(url: "https://specs.apollo.dev/join/v0.2", for: EXECUTION)
+        @link(url: "https://specs.apollo.dev/foo/v1.0", import: ["@foo"])
+
+        directive @foo(name: String!) on FIELD_DEFINITION
+
+        type Query {
+          a: User
+        }
+
+        type User @key(fields: "id") {
+          id: Int
+          name: String @foo(name: "graphA")
+        }
+      `,
+    };
+
+    const subgraphB = {
+      name: 'subgraphB',
+      typeDefs: gql`
+      extend schema
+        @link(url: "https://specs.apollo.dev/link/v1.0")
+        @link(url: "https://specs.apollo.dev/join/v0.2", for: EXECUTION)
+        @link(url: "https://specs.apollo.dev/foo/v1.0", import: ["@foo"])
+
+        directive @foo(name: String!) on FIELD_DEFINITION
+
+        type User @key(fields: "id") {
+          id: Int
+          description: String @foo(name: "graphB")
+        }
+      `,
+    };
+    const result = composeServices([subgraphA, subgraphB]);
+    expect(result.errors).toBeUndefined();
+    expect(result.hints?.length).toBe(0);
+    const { schema } = result;
+    expect(schema).toBeDefined();
+    if (schema) {
+      expectNoDirectiveOnElement(schema, 'User.name', 'foo');
+    }
+  });
+
+  it('type-system directive, core feature, explicitly exposed', () => {
+    const subgraphA = {
+      name: 'subgraphA',
+      typeDefs: gql`
+      extend schema
+        @link(url: "https://specs.apollo.dev/link/v1.0")
+        @link(url: "https://specs.apollo.dev/join/v0.2", for: EXECUTION)
+        @link(url: "https://specs.apollo.dev/foo/v1.0", import: ["@foo"])
+
+        directive @foo(name: String!) on FIELD_DEFINITION
+
+        type Query {
+          a: User
+        }
+
+        type User @key(fields: "id") {
+          id: Int
+          name: String @foo(name: "graphA")
+        }
+      `,
+    };
+
+    const subgraphB = {
+      name: 'subgraphB',
+      typeDefs: gql`
+      extend schema
+        @link(url: "https://specs.apollo.dev/link/v1.0")
+        @link(url: "https://specs.apollo.dev/join/v0.2", for: EXECUTION)
+        @link(url: "https://specs.apollo.dev/foo/v1.0", import: ["@foo"])
+
+        directive @foo(name: String!) on FIELD_DEFINITION
+
+        type User @key(fields: "id") {
+          id: Int
+          description: String @foo(name: "graphB")
+        }
+      `,
+    };
+    const result = composeServices([subgraphA, subgraphB], { exposeDirectives: ['@foo']});
+    expect(result.errors).toBeUndefined();
+    expect(result.hints?.length).toBe(0);
+    const { schema } = result;
+    expect(schema).toBeDefined();
+    if (schema) {
+      expectDirectiveOnElement(schema, 'User.name', 'foo', { name: 'graphA'});
+    }
+  });
+
+  it.todo('expose builtin directive');
```

Comment:
Worst adding that one :). Especially to be clear on what behaviour we want: should it be essentially ignored (since built-ins are merged by default) or should it error out (to make it clear the user is doing something not useful).

As I've hinted a bunch of time, I tend to like defaulting on erroring when something appears to be a user "mistake". My reasoning being that:
1. as a user, if I do something  not useful, I'm usually more happy than mad to learn about it, even if it's not harmful per-se.
2. I usually feel that if it's a choice between maybe preventing a mistake and maybe getting in the way, I'd rather risk getting in the way but prevent a mistake, because the cost of mistake tends to be much higher than small get-in-the-ways. It's a balance of course but ...
3. it's easier to remove a restriction than add one later.

Reply:
Agreed. Trying to promote builtin directives will error. That was the outstanding work I had.